pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java

```diff
@@ -93,5 +100,25 @@ protected void sendBlock(SendingMailbox<TransferableBlock> sendingMailbox, Trans
     }
   }
 
-  protected abstract void route(List<SendingMailbox<TransferableBlock>> destinations, TransferableBlock block);
+  protected abstract void route(List<SendingMailbox<TransferableBlock>> destinations, TransferableBlock block)
+      throws Exception;
+
+  // Called when the OpChain gracefully returns.
+  // TODO: This is a no-op right now.
+  public void close() {
+    for (SendingMailbox sendingMailbox : _sendingMailboxes) {
+      if (sendingMailbox.isInitialized() && !sendingMailbox.isClosed()) {
+        LOGGER.info("SendingMailbox={} was not closed presumably because receiver hasn't completed processing",
+            sendingMailbox.getMailboxId());
+      }
+    }
+  }
+
+  public void cancel(Throwable t) {
+    for (SendingMailbox sendingMailbox : _sendingMailboxes) {
+      if (sendingMailbox.isInitialized() && !sendingMailbox.isClosed()) {
+        sendingMailbox.cancel(t);
+      }
+    }
+  }
```

Comment:
for symmetric point of view. we can also add a "releaseSendingMailbox" method in MailboxService,

Reply:
There's no release for that since the ownership is not with `MailboxService`. By design the sending/receiving mailbox are asymmetric.